File: Java/139. Word Break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int[] dp = new int[s.length() + 1];
        dp[0] = 1;
        Set<String> wordSet = new HashSet<String>(wordDict);
        for (int i = 1; i < s.length() + 1; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] == 1 && wordSet.contains(s.substring(j, i)))
                    dp[i] = 1;
            }            
        }
        return dp[s.length()] == 1;
    }
}

File: Python/139. Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set(wordDict)
        wordMark = [False] * (len(s) + 1)
        wordMark[0] = True
        for i in range(1, len(s) + 1):
            for j in range(0, i):
                if wordMark[j] and s[j:i] in wordSet:
                    wordMark[i] = True
                    break
        return wordMark[len(s)]
